crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,140 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - (starred_index + 1);
+                                // This index represents the type of the last element that belongs
+                                // to the starred expression, in an exclusive manner.
+                                let starred_end_index = tuple_ty.len(builder.db) - remaining;
+                                // SAFETY: Safe because of the length check above.
+                                let _starred_element_types = tuple_ty
+                                    .elements(builder.db)
+                                    .get(starred_index..starred_end_index)
+                                    .unwrap();
+                                // TODO: Combine the types into a list type. If the
+                                // starred_element_types is empty, then it should be `List[Any]`.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
+
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty
+                                        .elements(builder.db)
+                                        .get(starred_end_index..)
+                                        .unwrap(),
+                                );
+                                Cow::Owned(element_types)
+                            } else {
+                                Cow::Borrowed(tuple_ty.elements(builder.db).as_ref())
+                            }
+                        } else {
+                            Cow::Borrowed(tuple_ty.elements(builder.db).as_ref())
+                        };
+
+                        for (index, element) in elts.iter().enumerate() {
+                            if let Some(ty) = inner(
+                                builder,
+                                element,
+                                element_types.get(index).copied().unwrap_or(Type::Unknown),
+                                variable,
+                            ) {
+                                return Some(ty);
+                            }
+                        }
+                    }
+                    Type::StringLiteral(string_literal_ty) => {
+                        // Deconstruct the string literal to delegate the inference back to the
+                        // tuple type for correct handling of starred expressions.
+                        let value_ty = Type::Tuple(TupleType::new(
+                            builder.db,
+                            vec![Type::LiteralString; string_literal_ty.len(builder.db)]
+                                .into_boxed_slice(),
+                        ));
```

Comment:
I think ideally, for something like this:

```py
a, (b,), c, d = "abcd"
```

We'd infer `a` as being `Literal["a"]`, `b` as `Literal["b"]`, `c` as `Literal["c"]` and `d` as `Literal["d"]`. (At the moment they're all `LiteralString`.)

Reply:
Yeah, I think that depends on how precise we want the types. For this case, I think it might make sense to do so.

Reply:
I can do that as a follow-up.

Reply:
Sounds good!

Reply:
I agree that we _can_ do this, I'm not sure we ever _need_ to do it (pyright doesn't do it, it will have some cost in complexity and perf), and I'm pretty sure there's no need to do it _now_. I would suggest we defer it until or unless we see real motivating use cases. Avoiding duplicate diagnostics, supporting unpacking in for loops, and handling starred elements correctly are all much more important follow-ups. (Though handling starred elements correctly is blocked on generic list type, I think.)

Reply:
fine by me if we don't do it, that all makes sense. but in that case I think we should leave a comment saying that it's a deliberate choice to infer a less precise type -- it _looked_ like it could have been a mistake to me when I was reading through the diff :-)

Reply:
Makes sense, I'll add such a comment!